File: 151_Reverse_Words_in_a_String.cpp
#include <iostream>
#include <cstdio>
#include <string>
#include <stack>
using namespace std;

class Solution {
public:
    string reverseWords(string s) {
        string result;
        stack <char> tempst;
        int i = s.length() - 1;
        while (i >= 0) {
            if (s[i] == ' '  && !tempst.empty()) {
                    while (!tempst.empty()){
                        result += tempst.top();
                        tempst.pop();
                    }
                    result += " ";
                }
            else if (s[i] == ' '  && tempst.empty()) {
                i--;
                continue;
            }
            else {
                tempst.push(s[i]);
            }
            i--;
        }    

        while (!tempst.empty()) {
            result += tempst.top();
            tempst.pop();
        }
        while (result.back() == ' '){
            result.pop_back();
        }
        return result;
    }
};

int main(){
    Solution sol;
    string s = "  hello world  ";
    cout << "||" <<sol.reverseWords(s) << "||" << endl;
    return 0;
}

File: 151_Reverse_Words_in_a_String.py
class Solution:
    def reverseWords(self, s: str) -> str:
        p = len(s) - 1
        stack = []
        word = ""
        while p >= 0:
            if s[p] == " " and not stack:
                p-=1
                continue
            elif s[p] == " " and stack:
                
                while stack:
                    word += stack.pop()
                word += " "
            else:
                stack.append(s[p])
            p-=1

        while stack:
            word += stack.pop()
        if word[-1] == " ":
            word = word[:-1]
        return word
            